examples/express/index.ts

```diff
@@ -0,0 +1,170 @@
+import * as dotenv from 'dotenv'
+import crypto from 'crypto'
+import cookieParser from 'cookie-parser'
+import open from 'open'
+
+dotenv.config()
+
+const PORT = 5001
+const redirectUri = String(
+  process.env.SGID_REDIRECT_URI ?? `http://localhost:${PORT}/api/callback`,
+)
+const frontendHost = String(
+  process.env.SGID_FRONTEND_HOST ?? 'http://localhost:5173',
+)
+
+const sgid = new SgidClient({
+  clientId: String(process.env.SGID_CLIENT_ID),
+  clientSecret: String(process.env.SGID_CLIENT_SECRET),
+  privateKey: String(process.env.SGID_PRIVATE_KEY),
+  redirectUri,
+})
+
+const app = express()
+
+const apiRouter = Router()
+
+const SESSION_COOKIE_NAME = 'exampleAppSession'
+const SESSION_COOKIE_OPTIONS = {
+  httpOnly: true,
+}
+
+type SessionData = Record<
+  string,
+  | {
+      nonce?: string
+      // Store state as search params to easily stringify key-value pairs
+      state?: URLSearchParams
+      accessToken?: string
+      codeVerifier?: string
+      sub?: string
+    }
+  | undefined
+>
+
+/**
+ * In-memory store for session data.
+ * In a real application, this would be a database.
+ */
+const sessionData: SessionData = {}
+
+app.use(
+  cors({
+    credentials: true,
+    origin: frontendHost,
+  }),
+)
+
+apiRouter.get('/auth-url', (req, res) => {
+  const iceCreamSelection = String(req.query.icecream)
+  const sessionId = crypto.randomUUID()
+  // Use search params to store state so other key-value pairs
+  // can be added easily
+  const state = new URLSearchParams({
+    icecream: iceCreamSelection,
+  })
+
+  // Generate a PKCE pair
+  const { codeChallenge, codeVerifier } = generatePkcePair()
+
+  const { url, nonce } = sgid.authorizationUrl({
+    // We pass the user's ice cream preference as the state,
+    // so after they log in, we can display it together with the
+    // other user info.
+    state: state.toString(),
+    codeChallenge,
+    // Scopes that all sgID relying parties can access by default
+    scope: ['openid', 'myinfo.name'],
+  })
+  sessionData[sessionId] = {
+    state,
+    nonce,
+    codeVerifier,
+  }
+  return res
+    .cookie(SESSION_COOKIE_NAME, sessionId, SESSION_COOKIE_OPTIONS)
+    .json({ url })
+})
+
+apiRouter.get('/callback', async (req, res): Promise<void> => {
+  const authCode = String(req.query.code)
+  const state = String(req.query.state)
+  const sessionId = String(req.cookies[SESSION_COOKIE_NAME])
+
+  const session = { ...sessionData[sessionId] }
+  // Validate that the state matches what we passed to sgID for this session
+  if (session?.state?.toString() !== state) {
+    res.redirect(`${frontendHost}/error`)
+    return
+  }
+
+  // Validate that the code verifier exists for this session
+  if (!session?.codeVerifier) {
+    res.redirect(`${frontendHost}/error`)
+    return
+  }
+
+  // Exchange the authorization code and code verifier for the access token
+  const { accessToken, sub } = await sgid.callback({
+    code: authCode,
+    nonce: session.nonce,
```

Comment:
nit: should we destructure the `nonce` and `codeVerifier` from the session? because in`/userinfo` the values are being extracted from the session before being passed to `userinfo` (might also make passing the params a bit neater)